### ISSUE TITLE
Before and After Step Design

### ISSUE THREAD

**Issue by pawelsubko** [ppwwyyxx/tensorpack#39]
Hi,
Could you please alow acces to "before_step" and "after_step" ? sometimes it is crucial to 
1. run something more often then once per epoch (or just with different frequency) 
2. to run something explicitely BEFORE train_op (like some tensor statistics)
3. it makes life easier defining timers etc. (it is crucial to benchmark times of train_op, other_run_ops, loading data etc.)

**Comment by ppwwyyxx** [comment]
Sure it won't be hard to make the changes. But meanwhile I'm also curious about what your use case is and see if there's a better solution for that.
1. To run things more frequently you can change the epoch size, or use `PeriodicCallback` to adjust the frequency. Unless you really want something to run very frequently (every several iterations), and I wonder what that could be.

2. What do you mean by tensor statistics? `before the train_op` seems no different from `after the last train_op`, unless you're using the less efficient SimpleTrainer which loads data between two calls to the op.

3. As said above, you cannot accurately test the time of data loading by this two callbacks, because it is supposed to run in parallel to train_op. 

To test the speed of your dataflow, you can either write a small function which simply loops over the dataflow, or do `TestDataSpeed(dataflow, size=1000).start_test()` which does the same thing.

The iteration speed given in training should be the speed of `train_op`, unless the data is slower, in which case you can use `FakeData` instead to obtain a speed measurement of `train_op`. And what do you mean by `other_run_ops`?

**Comment by pawelsubko** [comment]
0. That frequently? Just um... (it's complicated) sending some data to the server. Sending it every step is not the best solution.
I. Ok, I may completely miss the point. Could you help me on the following problem:
What is a best way to compute loss on a given datapoint, then run train_op on this datapoint? (otherwise it's a waste!)
II. The way I "normally" do is I benchmark the whole loop body including the data_generator.next() call and all the other calls, and the train step only to get reliable (?) benchmark whether GPU train_op or some other calls/cpu<->gpu transfers etc. are the bottleneck. (Last time inefficiently written Augmentation (?) caused 1s whole loop vs 450ms train_op, the problem may be somewhere else tho)

**Comment by ppwwyyxx** [comment]
I. I'm still not sure.. do you mean "if the loss is small enough, then don't train on this data point"?
If that is the case, you can probably just implement this in the graph, with tf.cond and tf.stop_gradient. But  maybe doing something on dataflow to not feed such datapoint would be nicer.

II. So you are using SimpleTrainer. And then you can get a reliable benchmark of train_op and data from the loop body. But this kind of benchmark won't work for any other queue-based trainers where data and train_op happen in separate threads.
Also, dataflow or tensorflow might need some time to warm up, so I'd just use `TestDataSpeed` to test the data and `FakeData` to test the graph, for a couple of hundred iterations.

**Comment by pawelsubko** [comment]
I. Given a datapoint I want to compute the loss of my model on this datapoint and make sure it is computed before the model was trained on this datapoint (in this epoch at least). Then, since the datapoint is already in RAM, I want to use it for training.

II. I like this way of benchmarking dataloading and graph. I find one major problem tho:
It is not real time while training (might be important for many reasons for a cluster with multi-GPU setup where resources are shared - it's a real case scenario and I need to know if my 7 days working task is lagging for some reason). Therefore at least some timing of body loop is a must.
Again - I may define shorter epoch (it may last up to 12 hours in some experiments), but becomes quite painful and not very convenient workaround.

**Comment by ppwwyyxx** [comment]
I. It is definitely already computed in that way. But currently the only way to access this information is through the graph (or summary). I was planning to allow train_op to support some output tensors (and accessible through `trigger_step()`).

II. For queue-based trainer, the timing **is** done of the body loop. It's just in a multi-queue pipeline, there is no way to accurately benchmark each component while the whole pipeline is running, because every component waits for others. I can add some statistics of, for example, the occupancy of the queues which reveals which component is the bottleneck. This is also used in some of the tensorflow official examples.

**Comment by pawelsubko** [comment]
I. Great.
II. I get what you mean. These statistics (relative time of occupancy or something like that) would be great. Although, the time of backpropagation alone (not matter if it's waiting for data or not) may still give you an Idea of how long other Callbacks take (sometimes it's as long as the backpropagation step alone).

Regarding the design I'm not saying it's particularly bad but now to use my own timer (it sends statistics somewhere) I have to override the context manager - not sure if that's what's desirable.

**Comment by ppwwyyxx** [comment]
I. You can just use self.cost in the graph however you like (print it with tf.Print, skip training conditioned on it (with tf.cond and tf.stop_gradient). It __is__ the cost of the current datapoint before backprop happens.
Maybe I lost some context here, but you don't need trigger_step to do this.
If you want to write some Python logic based on the loss, now you can do it by the new Callback interface with trigger_step:
```python
def _extra_fetches(self):
    return ['name of the cost tensor:0']
def _trigger_step(self, cost):
    print(cost)
```

II. (i) The queue occupancy statistics was added for quite a while.
(ii) Do you mean you want the time of the minimization op only, but don't want others?
For efficiency, the other ops that you need in each step (like the callbacks I drafted above) are run together with the minimization op in one sess.run call, so the time alone cannot be measured.
The time of the sess.run call and how you process the data (the trigger_step method) can be measured separately. Is that really necessary? I'm expecting trigger_step should never do anything heavy, and you can tell whether something is heavy or not from the training speed already.

III. Now you can use your own timer in trigger_step. I even implemented the [progress bar](https://github.com/ppwwyyxx/tensorpack/blob/a51e2de463d7b9bf62618e46c3c14529344f8dc3/tensorpack/callbacks/steps.py#L71) as a callback (which means you can use a different progress bar). But as I said, you'll only be able to measure the total time of running the graph + running all the trigger_step from all callbacks.

**Comment by pawelsubko** [comment]
I. That's great.
II. Not necessary. I just found benchmarking on the fly any given parts of the pipeline useful.
III. I'll definitely use a modified version of it. thanks.
